### ISSUE TITLE
Rolling back a transaction to a savepoint can hang if setFetchCount has been used

### ISSUE THREAD

**Issue by mbalmeida** [impossibl/pgjdbc-ng#538]
Hi,

I'm facing a few issues when rolling back a transaction if setFetchCount has been used (attached two test cases that demonstrate the main issues). More details on how to run the test cases is below.

I tested against stock PG 9.5 (using pgjdbc-ng 0.8.7). I have the impression the same happens with later PG releases but haven't confirmed yet.

Problem 1) rollback(savepoint) hangs if setFetchCount is used. Code snippet:

```
        connection.setAutoCommit(false);

        Savepoint savepoint = connection.setSavepoint();

        // Force the query to fail at runtime
        try (PreparedStatement preparedStatement =
                     connection.prepareStatement("SELECT * from t1 where i = (select i from t2);"))
        {
            preparedStatement.setFetchSize(1000);

            try (ResultSet set = preparedStatement.executeQuery()) {
                while (set.next()) {
                    int discarded = set.getInt(1);
                }
            }

        } catch (Exception e) {
            // This hangs!
            connection.rollback(savepoint);
       }
       
```

Problem 2) If I don't use a savepoint, rolling back the transaction causes a nullptr exception:

```
        connection.setAutoCommit(false);

        // Force the query to fail at runtime
        try (PreparedStatement preparedStatement =
                     connection.prepareStatement("SELECT * from t1 where i = (select i from t2);"))
        {
            preparedStatement.setFetchSize(1000);

            try (ResultSet set = preparedStatement.executeQuery()) {
                while (set.next()) {
                    int discarded = set.getInt(1);
                }
            }

        } catch (Exception e) {
            // This throws a nullptr exception
            connection.rollback();
       }
```

The above throws:

```
java.lang.NullPointerException
	at com.impossibl.postgres.system.BasicContext.query(BasicContext.java:625)
	at com.impossibl.postgres.jdbc.PGDirectConnection.lambda$execute$2(PGDirectConnection.java:566)
	at com.impossibl.postgres.jdbc.PGDirectConnection.lambda$execute$1(PGDirectConnection.java:464)
	at com.impossibl.postgres.jdbc.PGDirectConnection.execute(PGDirectConnection.java:490)
	at com.impossibl.postgres.jdbc.PGDirectConnection.execute(PGDirectConnection.java:463)
	at com.impossibl.postgres.jdbc.PGDirectConnection.execute(PGDirectConnection.java:566)
	at com.impossibl.postgres.jdbc.PGDirectConnection.rollback(PGDirectConnection.java:871)
	at Main.performRead1(Main.java:66)
	at Main.main(Main.java:120)
```

Happy to log these issues separately as well if you prefer.

Attached a maven based project that reproduces the issue. Usage (after unzipping and cd'ing into directory). The simple program receives 5 arguments where the first 4 represent the connection parameters.

```
mvn clean install
# The command below should hang
java -jar ./target/testRollback.jar hostname username password dbname 0

#The command below should throw the nullptr exception
java -jar ./target/testRollback.jar hostname username password dbname 1
```

[testDBDriver.zip](https://github.com/impossibl/pgjdbc-ng/files/6397887/testDBDriver.zip)

Regards,
Matheus

**Comment by kdubb** [comment]
@mbalmeida Thank you for your help👍 

The simple reproducers helped immensely.  I've fixed the issue and added the reproducers to the test suite. It will be included in the next release, which will be `0.8.8`.

**Comment by kdubb** [comment]
FYI, the fixed version is currently available in `0.9-SNAPSHOT`.
